### DISCUSSION TITLE
What is the preferred way to share a lint between Typescript and Tsx?

### DISCUSSION THREAD

**Discussion by TheAndrewJackson** [ast-grep/ast-grep#1519] [General]
I'd like to use the same rule in both Typescipt and Tsx files. I have a use case where the rule is compatible with both syntaxes. Something like this:

```yml
rule:
  any:
    - pattern: console.log($A)
fix:
  console.debug($A)
```

From what I can tell there are a few potential ways to do this. 

- Use a [global utility rule](https://ast-grep.github.io/guide/rule-config/utility-rule.html#global-utility-rules) and create a unique lint for for both Typescript and Tsx that use the utility
- Use [language injection](https://ast-grep.github.io/advanced/language-injection.html)
- Use [language globs](https://ast-grep.github.io/reference/sgconfig.html#languageglobs)

I got it work with language globs using this:

```yml
languageGlobs:
  Typescript: ['*.ts', '*.tsx']
```

I'd like to do something like this to avoid having to duplicate all of my lints with utility rules. I did see this warning in the docs though:
```
TypeScript and TSX are different.

TypeScript is a typed JavaScript extension and TSX is a further extension that allows JSX elements. They need different parsers because of [conflicting syntax](https://www.typescriptlang.org/docs/handbook/jsx.html#the-as-operator).

TS allows both the as operator and angle brackets (<>) for type assertions. While TSX only allows the as operator because it interprets angle brackets as JSX elements.
```

Is this the canonical way to configure this? Is this safe to do? Are there hidden pitfalls?

**Comment by HerringtonDarkholme** [comment]
At the moment there is no way to reuse rules between different languages by ast-grep.

>  Is this safe to do? Are there hidden pitfalls?

The issue here is that TSX and TS can have different pattern parsing and different kind. They are different in nature and just happen to share a lot of kinds.

A solution here is to use external tools to generate YAML files from template, for now.

**Reply by TheAndrewJackson** [reply]
Thanks! That makes sense. It's hard to support this when these ASTs aren't 100% compatible with each other.

I think what you're saying is some of the time a rule can match as expected in both file types but it may vary unexpectedly so you can't offer any concrete guarantees.

Apologies if this is obvious but I'm new to ASTs. Would it be possible to do a language injection with these files types? Similar to what's described in the docs for JS and CSS within HTML or CSS-in-JS. It's okay if this isn't possible right now. I'm mainly curious if it's technically possible or if there's something preventing it.
